Feat/supabase (#57)
teamhanko/docs: guides/databases/supabase.mdx

---
sidebarTitle: Supabase
title: Using Hanko with Supabase

> description: This guide showcases how to add new user info, like user ID and email to the Supabase Database, when they sign up. Jump to [Nextjs guide](/quickstarts/fullstack/next) to learn how to add Hanko Auth in your Next App.

Reviewer: ```suggestion
description: This guide showcases how to add new user info, like user ID and email to the Supabase database when  users sign up with Hanko. Jump to [Nextjs guide](/quickstarts/fullstack/next) to learn how to add Hanko Auth in your Next App.
```

---

> We're using Next.js in this guide, but don't worry, you can apply the same concepts into other frameworks or languages. We're working on adding examples with different frameworks, so stay tuned!

Reviewer: ```suggestion
We're using Next.js in this guide, but don't worry, you can apply the same concepts to other frameworks or languages. We're working on adding examples with different frameworks, so stay tuned!
```

<Steps>
  <Step title="Create an account on Supabase">
    Create an account on [Supabase](https://supabase.com/) if you haven't yet.
  </Step>
  <Step title="Create a new project">
    Once you've created an account, you'll be redirected to the dashboard. Click
    on 'New Project' to create a new project.
    <Frame>
      <img
        src="/images/supabase/create-new-project.png"
        alt="create new project"
        width="500"
        style={{ borderRadius: "0.5rem" }}
      />
    </Frame>
  </Step>
  <Step title="Create users table">

>     Now, in the project dashboard, choose 'Database' and create a new 'users' table.

Reviewer: ```suggestion
    Now, on the project dashboard, choose 'Database' and create a new 'users' table.
```

    We'll be using this table to add new user info, on signup. Along with the

>     default 'id' column, add 'user_id' and 'email' columns.

Reviewer: ```suggestion
    We'll be using this table to add new user info on signup. In addition to the
    default 'id' column, add 'user_id' and 'email' columns.
```

    <Frame>
      <img
        src="/images/supabase/create-new-table.png"
        alt="create new project"
        width="500"
        style={{ borderRadius: "0.5rem" }}
      />
    </Frame>
  </Step>
  <Step title="Get URL and service_role key from Supabase and add it to your project">

>     Now in the project dashboard, select 'Settings' from the left sidebar and then click on 'API'. Copy the 'URL' and 'service_role' key and paste them in your `.env.local` file.

Reviewer: ```suggestion
    Now in the project dashboard, select 'Project Settings' from the left sidebar and then click on 'API'. Copy the 'URL' and 'service_role' key and paste them in your `.env.local` file.
```

    ```sh .env.local
    NEXT_PUBLIC_HANKO_API_URL=https://f4****-4802-49ad-8e0b-3d3****ab32.hanko.io

    NEXT_PUBLIC_SUPABASE_PROJECT_URL=YOUR_SUPABASE_PROJECT_URL

>     NEXT_PUBLIC_SUPABASE_SERVICE_ROLE_KEY=YOUR_SUPABASE_SERVICE_ROLE_KEY

Reviewer: Is this the one in the `Project API keys` panel? The one marked with the red `secret` badge? The panel tells me to never share this publicly. Should this really be `NEXT_PUBLIC` then?

Author: Yeah, sorry. Actually, we don't need NEXT_PUBLIC in both project url and service role key

    ```

  </Step>
  <Step title="Install Supabase JavaScript Client">

>     Assuming you've already setup Hanko Auth, you'll need to install supabase to your to your app.

Reviewer: ```suggestion
    Assuming you've already set up Hanko Auth, you'll need to install supabase to your app.
```

    export async function POST(req: Request, res: Response) {
    try {
        const user = await req.json();

        if (!user) {
        return new NextResponse("Unauthorized", { status: 401 });
        }

        let supauser;

        try {
        const { data, error } = await supabase
            .from("users")
            .select("*")
            .eq("user_id", user.id)
            .single();

        if (error) {
            console.error(error);
        }

        supauser = data;

        if (!supauser) {
            const { data, error } = await supabase
            .from("users")
            .insert([{ user_id: user.id, email: user.email }]);

            if (error) {
            console.error(error);
            }

            supauser = data;
        }
        } catch (error) {
        console.error(error);
        }

        const user_data = {
        ...user,
        id: supauser?.id,
        };

        return new NextResponse(JSON.stringify(user_data), { status: 200 });
    } catch (error) {
        console.error("[CREATEUSER_ERROR]", error);
        return new NextResponse("Internal Error", { status: 500 });
    }

>     }

Reviewer: Indentation/Formatting for this code block is a bit off in a couple of places, looks weird when rendered.

    ```

  </Step>
  <Step title="Use the API in HankoAuth component">

>     Use the API to add a new user to 'users' table, as soon as user is authenticated and `onAuthFlowCompleted` is triggered.

Reviewer: ```suggestion
    Use the API to add a new user to the 'users' table, as soon as user is authenticated and `onAuthFlowCompleted` is triggered.
```

        );
    }, []);

    const redirectAfterLogin = useCallback(() => {
        // successfully logged in, redirect to a page in your application
        router.replace("/dashboard");
    }, [router]);

    useEffect(
        () =>
        hanko?.onAuthFlowCompleted(async () => {
            // here we use the api we created to add a new user to users table
            const user = await hanko.user.getCurrent();

            const fetchData = async () => {
            if (!user) {
                console.error("No user data");
                return;
            }
            try {
                const response = await fetch("/api/create-user", {
                method: "POST",
                body: JSON.stringify(user),
                });

                if (!response.ok)
                throw new Error(`HTTP error! status: ${response.status}`);
            } catch (error) {
                console.error("Fetch Error: ", error);
            }
            };
            await fetchData();
            redirectAfterLogin();
        }),
        [hanko, redirectAfterLogin]
    );

    useEffect(() => {
        register(hankoApi).catch((error) => {

>         // handle error

Reviewer: Why say "handle error" via comment here, but above errors are `console.error`-ed? I think it's best if things like this are consistent, so either `console.error` everywhere or comments everywhere.

        });
    }, []);

    return <hanko-auth />;

>     }

Reviewer: Indentation/Formatting off in some places.
